src/functions_framework/__init__.py

```diff
@@ -75,30 +97,47 @@ def _is_binary_cloud_event(request):
         and request.headers.get("ce-id")
     )
 
+# TODO(joelgerard): is this really legacy as I have stated?
+def _run_binary_legacy_cloud_event(function, request):
+    # Support CloudEvents in binary content mode, with data being the
+    # whole request body and context attributes retrieved from request
+    # headers.
+    data = request.get_data()
+    context = Context(
+        eventId=request.headers.get("ce-eventId"),
+        timestamp=request.headers.get("ce-timestamp"),
+        eventType=request.headers.get("ce-eventType"),
+        resource=request.headers.get("ce-resource"),
+    )
+    function(data, context)
+
+def _run_legacy_cloud_event(function, request):
+    # This is a regular CloudEvent
+    event_data = request.get_json()
+    if not event_data:
+        flask.abort(400)
+    event_object = _Event(**event_data)
+    data = event_object.data
+    context = Context(**event_object.context)
+    function(data, context)
+
+def _run_cloud_event(function, request, cloud_event_def):
+    event = _convert_request_to_event(request, cloud_event_def)
+    return function(event)
 
 def _event_view_func_wrapper(function, request):
     def view_func(path):
         if _is_binary_cloud_event(request):
-            # Support CloudEvents in binary content mode, with data being the
-            # whole request body and context attributes retrieved from request
-            # headers.
-            data = request.get_data()
-            context = Context(
-                eventId=request.headers.get("ce-eventId"),
-                timestamp=request.headers.get("ce-timestamp"),
-                eventType=request.headers.get("ce-eventType"),
-                resource=request.headers.get("ce-resource"),
-            )
-            function(data, context)
+            _run_binary_legacy_cloud_event(function, request)
         else:
-            # This is a regular CloudEvent
-            event_data = request.get_json()
-            if not event_data:
-                flask.abort(400)
-            event_object = _Event(**event_data)
-            data = event_object.data
-            context = Context(**event_object.context)
-            function(data, context)
+            cloud_event_def = _get_cloud_event_version(request)
+            if cloud_event_def is None:
+                _run_legacy_cloud_event(function, request)
+            else:
+                # We have a bonafide event from the Cloud Event SDK. Let's use it.
+                # TODO(joelgerard): Note that I return the function value. This seems better
+                # going forward rather than returning "OK". ??
+                return _run_cloud_event(function, request, cloud_event_def)
```

Comment:
I'm not sure we want the user to be able to modify what's being returned to the caller like this. Maybe an open question to discuss w/ the other framework implementations.

For example, right now it's OK for the user to return the implicit `None` from a function with the `event` signature type, and this will result a 200 response. For example, this function has no `return` statement:

https://github.com/GoogleCloudPlatform/functions-framework-python/blob/91afba29edf9ff4d7c3e08b50a2a47d61c9c5f10/tests/test_functions/background_trigger/main.py#L18-L36

If the user did the same thing here, though, it would cause an exception in the framework as it's only possible to return strings, tuples and `Reponse`-type objects. So if we did this, either the user need to start always returning something (and returning the right thing) like in the `http` events, or the framework needs to start introspecting the return value and modifying the response accordingly. I don't think we want either of those.